src/components/ic_drivers/tmc2209_driver/tmc2209_driver.c

```diff
@@ -0,0 +1,191 @@
+// Copyright 2024 MACKI, Krzysztof Gliwiński
+
+#include "tmc2209_driver.h"
+
+#include "tmc2209_datagram.h"
+#include "tmc2209_registers.h"
+
+#define TAG "TMC2209_DRIVER"
+
+#define TMC2209_STATUS_CHECK(status) \
+  if (status != TMC2209_DRIVER_OK) { \
+    return status;                   \
+  }
+
+#define TMC2209_WRITE_BIT 1
+#define TMC2209_READ_BIT 0
+#define TMC2209_WRITE_ACCESS_CONSTANT 0x80
+
+static tmc2209_driver_status_t _tmc2209_check_driver(tmc2209_driver_t *driver) {
+  if (driver == NULL) {
+    return TMC2209_DRIVER_ERROR;
+  }
+
+  return TMC2209_DRIVER_OK;
+}
+
+tmc2209_driver_status_t tmc2209_driver_init(tmc2209_driver_t *driver) {
+  if (driver == NULL) {
+    return TMC2209_DRIVER_ERROR;
+  }
+
+  tmc2209_reg_gstat_t gstat;
+  gstat.raw = 0;
+  gstat.drv_err = 1;
+
+  TMC2209_STATUS_CHECK(
+      tmc2209_write_register(driver, TMC_2209_REG_GSTAT, gstat.raw));
+
+  tmc2209_reg_gconf_t gconf;
+  gconf.raw = 0;
+  gconf.I_scale_analog = 0;
+  gconf.pdn_disable = 1;
+  gconf.mstep_reg_select = 1;
+  gconf.multistep_filt = 1;
+  tmc2209_reg_ifcnt_t ifcnt;
+
+  TMC2209_STATUS_CHECK(
+      tmc2209_write_register(driver, TMC_2209_REG_GCONF, gconf.raw));
+
+  TMC2209_STATUS_CHECK(tmc2209_driver_disable(driver));
+
+  TMC2209_STATUS_CHECK(
+      tmc2209_read_register(driver, TMC_2209_REG_GSTAT, &gstat.raw));
+  TMC2209_STATUS_CHECK(
+      tmc2209_write_register(driver, TMC_2209_REG_GSTAT, gstat.raw));
+
+  TMC2209_STATUS_CHECK(
+      tmc2209_read_register(driver, TMC_2209_REG_GCONF, &gconf.raw));
+
+  TMC2209_STATUS_CHECK(
+      tmc2209_read_register(driver, TMC_2209_REG_IFCNT, &ifcnt.raw));
+
+  uint8_t if_cnt = (uint8_t)ifcnt.raw;
+
+  gconf.pdn_disable = 1;
+  gconf.mstep_reg_select = 1;
+  TMC2209_STATUS_CHECK(
+      tmc2209_write_register(driver, TMC_2209_REG_GCONF, gconf.raw));
+
+  TMC2209_STATUS_CHECK(
+      tmc2209_read_register(driver, TMC_2209_REG_IFCNT, &ifcnt.raw));
+
+  tmc2209_driver_status_t ret = ((uint8_t)ifcnt.raw == if_cnt + 2)
+                                    ? TMC2209_DRIVER_OK
+                                    : TMC2209_DRIVER_ERROR;
+  if (ret != TMC2209_DRIVER_OK) {
+    return ret;
+  }
+  return TMC2209_DRIVER_OK;
+}
+
```

Comment:
Can you add set veliocity to 0 before enabling the driver, this should fix the problem with mottor running after esp reset.

Reply:
Added but as this driver is still non-functional will be uncommented later